runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/ThreadLocalByteStringOutputStream.java

```diff
@@ -0,0 +1,126 @@
+      ThreadLocal.withInitial(SoftRefHolder::new);
+
+  // Private constructor to prevent instantiations from outside.
+  private ThreadLocalByteStringOutputStream() {}
+
+  /** @return An AutoClosable StreamHandle that holds a cached ByteStringOutputStream. */
+  public static StreamHandle acquire() {
+    RefHolder refHolder = getRefHolderFromThreadLocal();
+    if (refHolder.inUse) {
+      // Stream is already in use, create a new uncached one
+      return new StreamHandle();
+    }
+    refHolder.inUse = true;
+    return Preconditions.checkArgumentNotNull(
+        refHolder.streamHandle); // inUse will be unset when streamHandle closes.
+  }
+
+  /**
+   * Handle to a thread-local {@link ByteStringOutputStream}. If the thread local stream is already
+   * in use, a new one is used. The streams are cached and reused across calls. Users should not
+   * keep a reference to the stream after closing the StreamHandle.
+   */
+  public static class StreamHandle implements AutoCloseable {
+
+    // When Nonnull the StreamHandle is from the threadlocal and needs to be
+    // marked as not in use, in close.
+    private final @Nullable RefHolder refHolder;
+    private final ByteStringOutputStream stream;
+
+    private StreamHandle() {
+      this.refHolder = null;
+      this.stream = new ByteStringOutputStream();
+    }
+
+    private StreamHandle(RefHolder refHolder) {
+      this.refHolder = refHolder;
+      this.stream = refHolder.stream;
+    }
+
+    /**
+     * Returns the underlying cached ByteStringOutputStream. Callers should not keep a reference to
+     * the stream after closing the StreamHandle.
+     */
+    public ByteStringOutputStream stream() {
+      return stream;
+    }
+
+    @Override
+    public void close() {
+      stream.reset();
+      if (refHolder != null) {
+        refHolder.inUse = false;
+      }
+    }
+  }
+
+  private static class SoftRefHolder {
+
+    private @Nullable SoftReference<RefHolder> softReference;
+  }
+
+  private static class RefHolder {
+
+    public ByteStringOutputStream stream = new ByteStringOutputStream();
+
+    // Boolean is true when the thread local stream is already in use by the current thread.
+    // Used to avoid reusing the same stream from nested calls if any.
+    public boolean inUse = false;
+
+    public @Nullable StreamHandle streamHandle = null;
+
+    public static RefHolder create() {
+      RefHolder refHolder = new RefHolder();
+      refHolder.streamHandle = new StreamHandle(refHolder);
+      return refHolder;
+    }
+  }
+
+  private static RefHolder getRefHolderFromThreadLocal() {
+    // softRefHolder is only set by Threadlocal initializer and should not be null
+    SoftRefHolder softRefHolder =
+        Preconditions.checkArgumentNotNull(threadLocalSoftRefHolder.get());
+    RefHolder refHolder;
+    if (softRefHolder.softReference != null && softRefHolder.softReference.get() != null) {
```

Comment:
how about assigning result of softRefHolder.softReference.get() to intermediate variable instead of calling it twice.
It seems like the precondition could fail if it was gc'd between checks and has more synchronization overhead to do twice.

```
@Nullable RefHolder refHolder;
if (softRefHolder.softReference != null) {
  refHolder = softRefHolder.softReference.get();
}
if (refHolder == null) {
  ...
}
return refHolder;
```